setup_dev.py

```diff
@@ -0,0 +1,532 @@
+#!/usr/bin/env python
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Sets up the development tree with various kinds of defaults.
+
+**DANGER WILL ROBINSON: This is experimental and is basically a serialization
+of Stella's dev process. We aim to use this to standardize a uniform dev
+process for out of the gate work by core team members, but the instructions
+on the website remain the authoritative well-lit path.**
+
+This script optimizes the case of a single development tree and CMake build
+directory, taking care to use every trick we have to make the experience
+uniform and with every development-appropriate performance trick we know
+how to apply.
+
+Usage:
+
+  ./setup_dev.py
+
+This will:
+
+* Try to select an appropriate C compiler a linker, preferring clang and lld
+  if available (controllable via --cc, --cxx, and --lld).
+* Configure all parts of the project to use ccache if available (controllable
+  via --no-ccache).
+* Configure all projects to build in release mode with debug symbols and
+  asserts enabled (configurable via --config).
+* Configure all compiler components to build shared development libraries
+  instead of static (controllable via --llvm-shared).
+* Enable several optimizations for more efficient handling of debug data and
+  static archives that have the side effect of making artifacts bound to the
+  precise dev setup (controllable via --no-dev).
+* Enables flags for Python.
+* Write out a cheat_sheet.md file with more commands to do things.
```

Comment:
file name here should be kept in sync - "dev_cheatsheet.md"
```suggestion
* Write out a dev_cheatsheet.md file with more commands to do things.
```

also - how much of that file could be a page in https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/developer_overview.md or another checked-in file? (yes, many of those pages are due for cleanup/rewrites)